File: examples/starter-go/memory.go
// Copyright 2026 The Kaalm Authors. Licensed under the Apache License, Version 2.0.

package main

import (
	"encoding/json"
	"log"
	"os"
	"path/filepath"
	"sync"
)

// memoryDir is where the operator mounts spec.persistence, and so where state
// that must outlive the Pod belongs. It matters for hibernation: a woken agent
// is a brand new Pod, and anything held only in RAM is gone. Override with
// KAALM_MEMORY_DIR if the Agent sets a custom spec.persistence.mountPath.
const defaultMemoryDir = "/var/agent/memory"

// persistedState is everything the agent carries across a restart. It is
// deliberately small and rewritten whole: an agent's working set here is a
// dedup window and a per-user counter, not a database.
type persistedState struct {
	// Dedup maps a delivered messageId to the reply that was returned for it,
	// so a redelivery after a wake is answered from cache rather than
	// processed twice (runtime contract item 7).
	Dedup map[string]ResponseEnvelope `json:"dedup"`
	// Seen counts messages per userId. The starter agent uses it to show that
	// memory survived; a real agent would keep conversation state here.
	Seen map[string]int `json:"seen"`
}

// store persists agent state to the mounted volume. With no writable volume it
// degrades to memory only, which is correct for an agent that does not enable
// hibernation: the contract requires persistence only when hibernationEnabled
// is set, and that in turn requires a PVC.
type store struct {
	path string // empty when there is no usable volume

	mu    sync.Mutex
	state persistedState
}

func newStore(dir string) *store {
	s := &store{state: persistedState{
		Dedup: map[string]ResponseEnvelope{},
		Seen:  map[string]int{},
	}}
	if dir == "" {
		return s
	}
	// Probe rather than assume: persistence is optional, and an agent without a
	// PVC must still run instead of crash-looping on a missing directory.
	if err := os.MkdirAll(dir, 0o700); err != nil {
		log.Printf("memory: %s unusable (%v); continuing without persistence", dir, err)
		return s
	}
	s.path = filepath.Join(dir, "state.json")
	s.load()
	return s
}

// load reads prior state. A missing or corrupt file is not fatal: the agent
// starts empty rather than refusing to serve.
func (s *store) load() {
	raw, err := os.ReadFile(s.path)
	if err != nil {
		if !os.IsNotExist(err) {
			log.Printf("memory: reading %s: %v; starting empty", s.path, err)
		}
		return
	}
	var loaded persistedState
	if err := json.Unmarshal(raw, &loaded); err != nil {
		log.Printf("memory: %s is not readable state (%v); starting empty", s.path, err)
		return
	}
	if loaded.Dedup != nil {
		s.state.Dedup = loaded.Dedup
	}
	if loaded.Seen != nil {
		s.state.Seen = loaded.Seen
	}
	log.Printf("memory: recovered %d cached replies and %d known users from %s",
		len(s.state.Dedup), len(s.state.Seen), s.path)
}

// flush writes the current state. The caller holds the lock.
func (s *store) flush() {
	if s.path == "" {
		return
	}
	raw, err := json.Marshal(s.state)
	if err != nil {
		log.Printf("memory: encoding state: %v", err)
		return
	}
	// Write a temp file and rename, so a crash mid-write cannot leave a
	// half-written state file that fails to parse on the next start.
	tmp := s.path + ".tmp"
	if err := os.WriteFile(tmp, raw, 0o600); err != nil {
		log.Printf("memory: writing %s: %v", tmp, err)
		return
	}
	if err := os.Rename(tmp, s.path); err != nil {
		log.Printf("memory: replacing %s: %v", s.path, err)
	}
}

// recall returns the cached reply for a messageId, if this agent already
// answered it (possibly in a previous life, before hibernation).
func (s *store) recall(messageID string) (ResponseEnvelope, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	reply, ok := s.state.Dedup[messageID]
	return reply, ok
}

// remember records the reply for a messageId, trimming the oldest entries when
// the window grows past size.
func (s *store) remember(messageID string, reply ResponseEnvelope, size int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Dedup[messageID] = reply
	for len(s.state.Dedup) > size {
		for k := range s.state.Dedup { // map order is random, which is fine for eviction
			delete(s.state.Dedup, k)
			break
		}
	}
	s.flush()
}

// note counts a message from a user and returns the running total, so the
// agent can show that it remembers them across restarts.
func (s *store) note(userID string) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Seen[userID]++
	count := s.state.Seen[userID]
	s.flush()
	return count
}

File: images/agent-python/kaalm_agent/memory.py
# Copyright 2026 The Kaalm Authors. Licensed under the Apache License, Version 2.0.
"""Persistent agent state on the mounted volume (runtime contract item 7).

A port of the Go starter's store semantics: probe the directory instead of
assuming it, degrade to memory-only when there is no usable volume, tolerate a
missing or corrupt state file, and write via temp-then-rename so a crash
mid-write cannot leave an unparseable file.

The one JSON file carries two disjoint areas: the runtime-owned dedup window
(messageId to cached reply, with LRU order), and a general key/value area.
Handler-visible state lives under the ``user/`` key prefix via UserMemory, so
handler keys can never collide with the dedup window.
"""

from __future__ import annotations

import json
import logging
import os
import threading
from pathlib import Path
from typing import Any

log = logging.getLogger("agent.memory")

DEDUP_CAP = 1024


class Store:
    """JSON-file-backed state with in-memory degradation."""

    def __init__(self, directory: str | None):
        self._lock = threading.Lock()
        self._kv: dict[str, Any] = {}
        self._dedup: dict[str, Any] = {}
        self._dedup_order: list[str] = []
        self._path: Path | None = None

        if not directory:
            return
        # Probe rather than assume: persistence is optional, and an agent
        # without a PVC must run instead of crash-looping on a missing dir.
        try:
            os.makedirs(directory, mode=0o700, exist_ok=True)
        except OSError as exc:
            log.warning("memory: %s unusable (%s); continuing without persistence", directory, exc)
            return
        self._path = Path(directory) / "state.json"
        self._load()

    @property
    def persistent(self) -> bool:
        return self._path is not None

    def _load(self) -> None:
        assert self._path is not None
        try:
            raw = self._path.read_text()
        except FileNotFoundError:
            return
        except OSError as exc:
            log.warning("memory: reading %s: %s; starting empty", self._path, exc)
            return
        try:
            loaded = json.loads(raw)
        except ValueError as exc:
            log.warning("memory: %s is not readable state (%s); starting empty", self._path, exc)
            return
        if not isinstance(loaded, dict):
            log.warning("memory: %s is not readable state (not an object); starting empty", self._path)
            return
        self._kv = dict(loaded.get("kv", {}))
        self._dedup = dict(loaded.get("dedup", {}))
        order = [m for m in loaded.get("dedupOrder", []) if m in self._dedup]
        # Re-derive order for any replies the order list lost track of.
        order.extend(m for m in self._dedup if m not in order)
        self._dedup_order = order
        log.info(
            "memory: recovered %d cached replies and %d keys from %s",
            len(self._dedup), len(self._kv), self._path,
        )

    def _flush_locked(self) -> None:
        if self._path is None:
            return
        raw = json.dumps({"kv": self._kv, "dedup": self._dedup, "dedupOrder": self._dedup_order})
        tmp = self._path.with_suffix(".json.tmp")
        try:
            tmp.write_text(raw)
            os.replace(tmp, self._path)
        except OSError as exc:
            log.warning("memory: writing %s: %s", self._path, exc)

    # ---- general key/value area ----

    def get(self, key: str, default: Any = None) -> Any:
        with self._lock:
            return self._kv.get(key, default)

    def put(self, key: str, value: Any) -> None:
        """Store a JSON-serializable value. Raises TypeError on anything else,
        loudly and immediately, rather than corrupting the state file later."""
        json.dumps(value)
        with self._lock:
            self._kv[key] = value
            self._flush_locked()

    def delete(self, key: str) -> None:
        with self._lock:
            self._kv.pop(key, None)
            self._flush_locked()

    # ---- runtime-owned dedup window (contract item 7) ----

    def recall(self, message_id: str) -> Any | None:
        """The cached reply for a messageId this agent already answered,
        possibly in a previous life, before hibernation."""
        with self._lock:
            reply = self._dedup.get(message_id)
            if reply is not None:
                self._dedup_order.remove(message_id)
                self._dedup_order.append(message_id)
            return reply

    def remember(self, message_id: str, reply: Any, cap: int = DEDUP_CAP) -> None:
        with self._lock:
            if message_id not in self._dedup:
                self._dedup_order.append(message_id)
            self._dedup[message_id] = reply
            while len(self._dedup_order) > cap:
                oldest = self._dedup_order.pop(0)
                self._dedup.pop(oldest, None)
            self._flush_locked()


class UserMemory:
    """The handler-facing view of the store: the same persistence and
    degradation semantics, confined to the ``user/`` prefix (the kaalm module
    ABI, docs/src/runtime/base-images.md)."""

    _PREFIX = "user/"

    def __init__(self, store: Store):
        self._store = store

    def get(self, key: str, default: Any = None) -> Any:
        return self._store.get(self._PREFIX + key, default)

    def put(self, key: str, value: Any) -> None:
        self._store.put(self._PREFIX + key, value)

    def delete(self, key: str) -> None:
        self._store.delete(self._PREFIX + key)

    @property
    def persistent(self) -> bool:
        return self._store.persistent
